src/ray/gcs/store_client/in_memory_store_client.cc

```diff
@@ -0,0 +1,173 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/gcs/store_client/in_memory_store_client.h"
+
+namespace ray {
+
+namespace gcs {
+
+void InMemoryTable::Put(const std::string &key, const std::string &data) {
+  absl::MutexLock lock(&mutex_);
+  records_[key] = data;
+}
+void InMemoryTable::PutWithIndex(const std::string &key, const std::string &index_key,
+                                 const std::string &data) {
+  absl::MutexLock lock(&mutex_);
+  records_[key] = data;
+  index_keys_[index_key].push_back(key);
+}
+
+boost::optional<std::string> InMemoryTable::Get(const std::string &key) {
+  absl::MutexLock lock(&mutex_);
+  if (records_.count(key)) {
+    return records_[key];
+  }
+  return boost::none;
+}
+
+std::unordered_map<std::string, std::string> InMemoryTable::GetAll() {
+  absl::MutexLock lock(&mutex_);
+  return records_;
+}
+
+void InMemoryTable::Delete(const std::string &key) {
+  absl::MutexLock lock(&mutex_);
+  records_.erase(key);
+}
+
+void InMemoryTable::DeleteByIndex(const std::string &index_key) {
+  absl::MutexLock lock(&mutex_);
+  auto delete_keys = index_keys_[index_key];
```

Comment:
It will query twice. Pls use find and erase to optimize it

Reply:
fixed